source/distributed/fully_distributed_tria_util.cc

```diff
@@ -0,0 +1,149 @@
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/geometry_info.h>
+#include <deal.II/base/mpi.h>
+
+#include <deal.II/distributed/fully_distributed_tria.h>
+#include <deal.II/distributed/fully_distributed_tria_util.h>
+#include <deal.II/distributed/tria.h>
+
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/grid/grid_tools.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace fullydistributed
+  {
+    namespace Utilities
+    {
+      namespace
+      {
+      template <int dim, int spacedim>
+      ConstructionData<dim, spacedim>
+      create_construction_data_from_triangulation_in_goups(
+        std::function<void(dealii::Triangulation<dim, spacedim> &)>
+                                                serial_grid_generator,
+        std::function<void(dealii::Triangulation<dim, spacedim> &,
+                           const MPI_Comm,
+                           const unsigned int)> serial_grid_partitioner,
+        const MPI_Comm                          comm,
+        const int                               group_size,
+        const bool                              construct_multilevel_hierarchy)
+      {
+        const unsigned int my_rank =
+          dealii::Utilities::MPI::this_mpi_process(comm);
+        const unsigned int group_root = (my_rank / group_size) * group_size;
+
+        if (my_rank == group_root)
+          {
+            // Step 1: create serial triangulation
+            dealii::Triangulation<dim, spacedim> tria(
+              construct_multilevel_hierarchy ?
+                dealii::Triangulation<dim, spacedim>::none :
+                dealii::Triangulation<dim, spacedim>::
+                  limit_level_difference_at_vertices);
+            serial_grid_generator(tria);
+
+            // Step 2: partition active cells and ...
+            serial_grid_partitioner(tria, comm, group_size);
+
+            // ... cells on the levels if multigrid is required
+            if (construct_multilevel_hierarchy)
+              GridTools::partition_multigrid_levels(tria);
+
+
+            const unsigned int max_rank =
+              std::min(group_root + group_size,
+                       dealii::Utilities::MPI::n_mpi_processes(comm));
+
+            for (unsigned int other_rank = group_root + 1;
+                 other_rank < max_rank;
+                 other_rank++)
+              {
+                // create construction data for other ranks
+                auto construction_data =
+                  create_construction_data_from_triangulation(
+                    tria, comm, construct_multilevel_hierarchy, other_rank);
+                // pack
+                std::string                                       serial_str;
+                boost::iostreams::back_insert_device<std::string> inserter(
```

Comment:
@kronbichler I guess I could use `Utilities::pack` and `Utilities::unpack` here?